### ISSUE TITLE
"WARNING: Retrying failed request" when listing objects in a bucket

### ISSUE THREAD

**Issue by matteobar** [s3tools/s3cmd#314]
When listing all the objects in a S3 bucket, the following WARNING is shown every time:

WARNING: Retrying failed request: /?marker=100PENTX/IMGP0125.JPG ()
WARNING: Waiting 3 sec...

In the end the listing does succeed, but this warning is given every time.
Looks like it's something to do with a failed "next marker" request, the bucket contains more than 1000 files. After initial failure, it always succeeds.

**Comment by allella** [comment]
FWIW, there are a few potential fixes/theories posted at
http://stackoverflow.com/questions/5774808/s3cmd-failed-too-many-times

**Comment by jamieburchell** [comment]
Thanks but in my case it's files of any size, they always need to be "retried" the first time an upload starts and they always get uploaded in the end (no fail)

**Comment by hrchu** [comment]
+1,
since multipart upload works, this is not an big problem for me.

**Comment by jamieburchell** [comment]
I've posted a sample log file showing the retries, the broken pipe and the eventual successful upload. This process seems to always happen regardless of the file being uploaded. It always uploads the second time round.

http://pastebin.com/raw.php?i=j3nQgNaR

**Comment by tharple** [comment]
Since installing the latest version today, (To use stdin functionality)
Comparing: s3cmd --version: 
s3cmd version 1.1.0-beta3 
to 
s3cmd version 1.5.0
ls without bucket name both pass.
ls with bucket name:
1.1.0-beta never fails.
1.5.0 always fails
observed debug difference(format_uri):

1.1.0-beta3: 

DEBUG: format_uri(): /?delimiter=/
DEBUG: Sending request method_string='GET', uri='/?delimiter=/', headers={'content-length': '0', 'Authorization': 'AWS 7d798fd9af074
834bca0db006f152c85:1QbELjS7SLmBonw0g3ahbptpXPA=', 'x-amz-date': 'Thu, 22 Jan 2015 20:05:39 +0000'}, body=(0 bytes)
DEBUG: Response: {'status': 200, 'headers': {'content-length': '307', 'expires': 'Thu, 01 Jan 1970 00:00:00 GMT', 'server': 'Verizon
-Himalaya/3.5.1-9e1b68cffea7', 'pragma': 'no-cache', 'cache-control': 'no-cache', 'date': 'Thu, 22 Jan 2015 20:05:37 GMT', 'content-
type': 'application/xml;charset=UTF-8'}, 'reason': 'OK', 'data': '<?xml version="1.0" encoding="UTF-8"?><ListBucketResult xmlns="htt
p://s3.amazonaws.com/doc/2006-03-01/"><Name>tharple-1</Name><Prefix></Prefix><Marker/><MaxKeys>1000</MaxKeys><Delimiter>/</Delimiter
><IsTruncated>false</IsTruncated><CommonPrefixes><Prefix>scalar/</Prefix></CommonPrefixes></ListBucketResult>'}

1.5.0:

DEBUG: format_uri(): /tharple-1/?delimiter=/
DEBUG: Sending request method_string='GET', uri='/tharple-1/?delimiter=/', headers={'x-amz-content-sha256': 'e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855', 'x-amz-date': '20150122T195452Z', 'Authorization': 'AWS4-HMAC-SHA256 Credential=7d798fd9af074834bca0db006f152c85/20150122/US/s3/aws4_request,SignedHeaders=host;x-amz-content-sha256;x-amz-date,Signature=c4424a556b5a77a84479aeecd0ef34fdc247c8a7301a79802a7cdc53f73426e1'}, body=(0 bytes)
DEBUG: Response: {'status': 500, 'headers': {'content-length': '194', 'dav': '1,3', 'server': 'Verizon-Himalaya/3.5.1-9e1b68cffea7', 'connection': 'close', 'date': 'Thu, 22 Jan 2015 19:55:37 GMT', 'content-type': 'application/xml'}, 'reason': 'Internal Server Error', 'data': '<?xml version="1.0" encoding="UTF-8"?>\n<Error><Code>InternalError</Code><Message>We encountered an internal error. Please try again.</Message><RequestId></RequestId><Resource></Resource></Error>'}

Hope this helps
Terry

**Comment by allella** [comment]
Thanks Terry. Is that your AWS key and secret in the debugging? If so, you should probably invalidate that key so some evildoer doesn't find this and access your bucket.

**Comment by mdomsch** [comment]
Verizon Cloud Storage (aka Himalaya)  is throwing an HTTP 500 error.  For
now, maybe try with --signature-v2 because it clearly doesn't like
something about the v4 signed requests.  This is also the first instance
I've heard of someone using s3cmd against Verizon Cloud Storage - glad to
know it has worked in the past.

Sounds like we need to get a test matrix built that includes all of the S3
clones (Swift, Walrus, DreamObjects, now Verizon Cloud Storage, fakes3 I
think I've used it with Google Cloud too...)  I definitely don't test
against all of them myself.

**Comment by tharple** [comment]
Thanks allella
Already removed

**Comment by mdomsch** [comment]
*Note: *DNS resolution for host-style addressing of buckets does not work
on buckets with names that contain a '.' (period). The same is true for SSL
certificate checking.

http://cloud.verizon.com/documentation/Buckets.htm


This means that we can't use s3cmd 1.5.0 against Verizon Cloud if your
bucket has a '.' in its name.  They are not handling wildcard NS or
subdomain A records in their DNS, so they won't resolve these.  s3cmd uses
https://<bucketname>.<endpointname>/path/to/object in its URLs, which won't
resolve.  It doesn't use https://<endpointname>/<bucketname>/path/to/object
anymore.

**Comment by jackal242** [comment]
Any updates on this bug?  I'm having the same problem.

I'm trying to run  "s3cmd ls --recursive s3://mybucket/90210/" and I'm getting a gazillion of these:

WARNING: Retrying failed request: /?marker=90210/lalala/FILESTUFF1 ()
WARNING: Waiting 3 sec...
WARNING: Retrying failed request: /?marker=90210/lalala/FILESTUFF2 ()
WARNING: Waiting 3 sec...

There are over a 1000 items in this list.

**Comment by foresto** [comment]
To anyone still struggling with this bug: I ran into it with s3tools version version 1.5.0~rc1 (which is the version in the Ubuntu 15.04 repositories) but it went away when I updated to version 1.5.2.

**Comment by jamieburchell** [comment]
I was all excited to see that, until I realised I'm already using 1.5.2 and still have a log full of them ;(

**Comment by fviard** [comment]
@jamieburchell: If you are consistently able to reproduce the issue.
Could it be possible for you to run the 1.5.2 with the "--debug" command line option, and then post the log? (take care to remove any personal sensitive data)
Thanks

**Comment by jamieburchell** [comment]
@fviard Does this one help?
https://github.com/s3tools/s3cmd/issues/485#issuecomment-76597331

**Comment by fviard** [comment]
@jamieburchell : Hum, could you do a similar log, but with the last upstream version of s3cmd? because in 1.5.2, there should already be the fixes that are described by @mdomsch. Thanks

**Comment by jamieburchell** [comment]
Please ignore my previous link to that debug - it was for a different issue. Take a look here (search for "retry" in the text)

http://jamieburchell.com/backup2s3.zip

Cheers

**Comment by jamieburchell** [comment]
That is a debug log from the 1.5.2 tag.

**Comment by fviard** [comment]
@jamieburchell Thanks for the log, that is really interesting.
A first thing that I noticed that could (or could not) be related to the issue is that your files are stored with the storage class "GLACIER". That probably could have an impact.

If you can try a few things, I think that we can narrow down the issue.
(It would be better if you can use the last master of s3cmd github but, it is still ok with the 1.5.2)
I would like that you run again your test in debug mode, but please tweak the following in your s3cmd code:

In S3/S3.py, in the following function:
def send_request(self, request, retries = _max_retries):
You will see the following line:
warning("Retrying failed request: %s (%s)" % (resource['uri'], e))
That would be great if you can add the following code just after:
'''
from logging import exception as log_exc
log_exc("WHY do I want to retry:")
'''

**Comment by jamieburchell** [comment]
I'll have a look in a bit. I have lifecycle rules setup to archive music to Glacier. This also happens on files that are not archived though.

**Comment by jamieburchell** [comment]
This extract from the master version:

```
ERROR: WHY do I want to retry:
Traceback (most recent call last):
  File "build/bdist.linux-armv5tel/egg/S3/S3.py", line 1014, in send_request
    http_response = conn.c.getresponse()
  File "/share/MD0_DATA/.qpkg/Python/lib/python2.7/httplib.py", line 1025, in getresponse
    response.begin()
  File "/share/MD0_DATA/.qpkg/Python/lib/python2.7/httplib.py", line 401, in begin
    version, status, reason = self._read_status()
  File "/share/MD0_DATA/.qpkg/Python/lib/python2.7/httplib.py", line 365, in _read_status
    raise BadStatusLine(line)
BadStatusLine: ''
```

**Comment by jamieburchell** [comment]
@fviard Any ideas?

**Comment by fviard** [comment]
@jamieburchell Sorry for the delay of the reply, it is more clear now, but i'm not sure that we can do something about this.
https://github.com/boto/boto/issues/1934
It looks like that somehow, when reusing connection in the case of keep alive, amazon send bad HTTP reply, like with extra spaces or things like that. httplib is unable to handle that and so fail.

But it is strange that we almost never encounter it whereas you face it often. So maybe, you can try to update httplib or the python of your device because they have improved something in the lib.
Otherwise, maybe we can retry with a shorter delay than 3s in such a case as the new connection will certainly fix the transfer for the request that encountered this issue.

**Comment by jamieburchell** [comment]
Hey

I'm not sure if I can update it, it's the version of Python that's available for the NAS and I think I looked into that before.

Cheers
Jamie

**Comment by jamieburchell** [comment]
OK I am running Python 2.7 which seems to be the latest version. s3cmd doesn't seem to work at all with Python 3

**Comment by jamieburchell** [comment]
Just to say these warnings on listing bucket files is still present in 1.0.6

```
INFO: Retrieving list of remote files for s3://xxx/Applications/ ...
WARNING: Retrying failed request: /Applications/Adobe+Acrobat+9/Adobe+Acrobat+9+Std.zip ('')
WARNING: Waiting 3 sec...
```

**Comment by connaryscott** [comment]
i have the same problem using lambda functions if the function call lasts longer than 1 minute.

```
ConnectionError: ('Connection aborted.', BadStatusLine("''",))
```

Only thing I can think of that may cause this is the use of a proxy.  When I use this in another environment without a proxy, this function call will last up to 5 minutes if needed (which is the maximum length or fime a lambda function may run if requested)

**Comment by fviard** [comment]
This issue should now be fixed in master.
Please give it a try.

**Comment by Joshfindit** [comment]
Just a quick one since this is the first issue I've had with `s3cmd`, and this issue is the first page I came across on the journey to eventually solving it.

The version available through Ubuntu's standard repositories is outdated. To update you can enable `Universe` (which I typically choose not to do on servers), or as per https://github.com/s3tools/s3cmd/blob/master/INSTALL you can do:

`sudo apt install python-pip`
followed by
`sudo pip install s3cmd`.

**Comment by fviard** [comment]
Should already be fixed now.

**Comment by fviard** [comment]
I'm encountering the issue now so it looks like that the issue is not fixed.
It looks like that it is exactly the same issue as:
https://github.com/boto/boto/issues/1934

Retrying a configure test multiple times I can encounter it and can confirm that there looks like to be a very short timeout with https keep alive timeout of aws.
If http is used, no problem. If https is used, reusing a connection more than 1s after the previous usage result in such error. At least s3cmd is able to recover, but it is bad for performance.

**Comment by fviard** [comment]
This issue should now be fixed with #1114 I think
